shell/platform/fuchsia/flutter/vsync_waiter.cc

```diff
@@ -61,14 +64,56 @@ VsyncWaiter::~VsyncWaiter() {
   ui_latch.Wait();
 }
 
-static fml::TimePoint SnapToNextPhase(fml::TimePoint value,
-                                      fml::TimePoint phase,
-                                      fml::TimeDelta interval) {
-  fml::TimeDelta offset = (phase - value) % interval;
-  if (offset < fml::TimeDelta::Zero()) {
-    offset = offset + interval;
+/// Returns the system time at which the next frame is likely to be presented.
+///
+/// Consider the following timelines, where in both the scenarious the result
+/// will be the same.
+///
+/// Scenario 1:
+/// presentation_interval is 2
+/// ^     ^     ^     ^     ^
+/// +     +     +     +     +
+/// 0--1--2--3--4--5--6--7--8--9--
+/// +        +  +
+/// |        |  +---------> result: next_presentation_time
+/// |        v
+/// v        now
+/// last_presentation_time
+///
+/// Scenario 2:
+/// presentation interVal is 2
+/// ^     ^     ^     ^     ^
+/// +     +     +     +     +
+/// 0--1--2--3--4--5--6--7--8--9--
+///       +  +  +
+///       |  |  +--------->result: next_presentation_time
+///       |  |
+///       |  +>now
+///       |
+///       +->last_presentation_time
+fml::TimePoint VsyncWaiter::SnapToNextPhase(
+    const fml::TimePoint now,
+    const fml::TimePoint last_frame_presentation_time,
+    const fml::TimeDelta presentation_interval) {
+  FML_CHECK(presentation_interval > fml::TimeDelta::Zero())
+      << "Presentation interval whould be positive.";
+  FML_DCHECK(last_frame_presentation_time < now)
+      << "Last frame was presented in the future. This is unexpected.";
+
+  const fml::TimeDelta delta = now - last_frame_presentation_time;
+  // this will be the most likely scenario if we are rendering at a good
+  // frame rate, short circuiting the other checks in this case.
+  if (delta < presentation_interval) {
+    return last_frame_presentation_time + presentation_interval;
+  } else {
+    int64_t num_phases_passed = delta / presentation_interval;
```

Comment:
unsigned?

Reply:
not sure it matters here since the overloaded division operator in `TimeDelta` is `int64_t` though as you point out it ought to be unsigned. Leaving this as is for now.